core/document/src/lib.rs

```diff
@@ -22,31 +26,148 @@ impl SvgElement {
 	}
 }
 
-#[derive(Default, Debug, Clone, PartialEq)]
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum DocumentError {
+	ElementNotFound,
+	NotAShape,
+	ElementAlreadyExists,
+	InvalidPath,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Folder {
+	elements: HashMap<String, SvgElement>,
+}
+
+impl Folder {
+	fn add_element(&mut self, name: String, svg: SvgElement) -> Result<(), DocumentError> {
+		#[allow(clippy::map_entry)]
+		if self.elements.contains_key(&name) {
+			Err(DocumentError::ElementAlreadyExists)
+		} else {
+			self.elements.insert(name, svg);
+			Ok(())
+		}
+	}
+
+	fn remove_element(&mut self, name: &str) -> Result<(), DocumentError> {
+		self.elements.remove(name).map_or(Err(DocumentError::ElementNotFound), |_| Ok(()))
+	}
+
+	/// Returns a list of elements in the folder
+	pub fn list(&self) -> Keys<String, SvgElement> {
+		self.elements.keys()
+	}
+
+	fn element(&self, name: &str) -> Option<&SvgElement> {
+		self.elements.get(name)
+	}
+
+	fn mut_element(&mut self, name: &str) -> Option<&mut SvgElement> {
+		self.elements.get_mut(name)
+	}
+}
+
+impl Default for Folder {
+	fn default() -> Self {
+		Self { elements: HashMap::new() }
+	}
+}
+
+#[derive(Debug, Clone, PartialEq)]
 pub struct Document {
-	pub svg: Vec<SvgElement>,
+	pub svg: SvgElement,
```

Comment:
Should the root actually be a folder, so it can contain N number of `SvgElements` inside?

Reply:
I'm not sure what you mean, the root is a folder

Reply:
I mean changing this line to say `pub svg: Folder,`